doc/lsst.daf.butler/datastores.rst

```diff
@@ -0,0 +1,79 @@
+This specifies the fully qualified class name of the Python class implementing the datastore.
+The default is to use the `~lsst.daf.butler.datastores.posixDatastore.PosixDatastore`.
+All other keys depend on the specific datastore class that is selected.
+
+The default configuration values can be inspected at ``$DAF_BUTLER_DIR/config`` and current values can be obtained by calling ``butler config-dump`` on a butler repository.
+
+File-Based Datastores
+=====================
+
+The file-based datastores (for example `~lsst.daf.butler.datastores.posixDatastore.PosixDatastore` and `~lsst.daf.butler.datastores.s3Datastore.S3Datastore`) share configuration since they use formatters to serialize datasets to file artifacts using file name template schemes and also support composite disassembly.
+
+The supported configurations are:
+
+root
+    The location of the "root" of the datastore "file system".
+    Usually the default value of ``<butlerRoot>/datastore`` can be left unchanged.
+    Here ``<butlerRoot>`` is a magic value that is replaced either with the location of the Butler configuration file or the top-level ``root`` as set in that ``butler.yaml`` configuration file.
+records
+    This sections defines the name of the registry table that should be used to hold details about datasets stored in the datastore (such as the path within the datastore and the associated formatter).
+    This only needs to be set if multiple datastores are to be used simultaneously within one Butler repository since the table names should not clash.
+create
+    A Boolean to define whether an attempt should be made to initialize the datastore by creating the directory.  Defaults to `True`.
+templates
+    The template to use to construct "files" within the datastore.
+    The template uses data dimensions to do this.
+    Generally the default setting will be usable although it can be tuned per `~lsst.daf.butler.DatasetType`, `~lsst.daf.butler.StorageClass` or data ID.
+formatters
+    Mapping of `~lsst.daf.butler.DatasetType`, `~lsst.daf.butler.StorageClass` or data ID to a specific formatter class that understands the associated Python type and will serialize it to a file artifact.
+    The formatters section also supports the definitions of write recipes (bulk configurations that can be selected for specific formatters) and write parameters (parameters that control how the dataset is serialized; note it is required that all serialized artifacts be readable by a formatter without knowing which write parameters were used).
+constraints
+    Specify `~lsst.daf.butler.DatasetType`, `~lsst.daf.butler.StorageClass` or data ID that will be accepted or rejected by this datastore.
+composites
+    Controls whether composite datasets are disassembled by the datastore.
+    By default composites are not disassembled.
+    Disassembly can be controlled by `~lsst.daf.butler.DatasetType`, `~lsst.daf.butler.StorageClass` or data ID.
```

Comment:
This is what the default looks like (it's really there as an example):
```yaml
# True: concrete composite (composites are disassembled);
# False: virtual composite (composites are written monolithically).
# default applies to all composites that are not explicitly mentioned
# elsewhere.
default: false
# Use a dict rather than a list to allow easy merging of user config
# into default config. Config class has no way of merging lists.
# Types can be StorageClass names or DatasetType names, with DatasetType
# name always taking preference over StorageClass name.
disassembled:
  ExposureI: false
  ExposureF: false
  calexp: false
```